core/templates/pages/diagnostic-test-player-page/diagnostic-test-topic-tracker.model.ts

```diff
@@ -0,0 +1,265 @@
+  // successor of B is only C.
+  _topicIdToSuccessorTopicIds: TopicIdToRelatedTopicIds;
+
+  constructor(topicIdToPrerequisiteTopicIds: TopicIdToRelatedTopicIds) {
+    this._failedTopicIds = [];
+    this._topicIdToAncestorTopicIds = {};
+    this._topicIdToSuccessorTopicIds = {};
+
+    this._topicIdToPrerequisiteTopicIds = topicIdToPrerequisiteTopicIds;
+    this._eligibleTopicIds = Object.keys(
+      this._topicIdToPrerequisiteTopicIds).sort();
+
+    this.generateTopicIdToAncestorTopicIds();
+    this.generateTopicIdToSuccessorTopicIds();
+  }
+
+  getEligibleTopicIds(): string[] {
+    return this._eligibleTopicIds;
+  }
+
+  getFailedTopicIds(): string[] {
+    return this._failedTopicIds;
+  }
+
+  getAncestorTopicIds(topicId: string): string[] {
+    return this._topicIdToAncestorTopicIds[topicId];
+  }
+
+  getTopicIdToAncestorTopicIds(): TopicIdToRelatedTopicIds {
+    return this._topicIdToAncestorTopicIds;
+  }
+
+  getSuccessorTopicIds(topicId: string): string[] {
+    return this._topicIdToSuccessorTopicIds[topicId];
+  }
+
+  getTopicIdToSuccessorTopicIds(): TopicIdToRelatedTopicIds {
+    return this._topicIdToSuccessorTopicIds;
+  }
+
+
+  generateTopicIdToAncestorTopicIds(): void {
+    // The method generates a dict with topic ID as the key and all of its
+    // ancestor topic IDs as value. The ancestor's list is generated using the
+    // prerequisite dependency graph.
+    // Example: A -> B -> C, here the topic ID to ancestor topic IDs dict will
+    // look like: {A: [], B: [A], C: [A, B]}.
+    for (let topicId in this._topicIdToPrerequisiteTopicIds) {
+      let ancestorTopicIds: string[] = [];
+      let unprocessedAncestorTopicIds: string[] = cloneDeep(
+        this._topicIdToPrerequisiteTopicIds[topicId]);
+
+      let visitedTopicIdsForCurrentTopic: string[] = [];
+      let lastTopicId: string;
+
+      while (unprocessedAncestorTopicIds.length > 0) {
+        lastTopicId = unprocessedAncestorTopicIds.pop() || '';
+
+        if (ancestorTopicIds.indexOf(lastTopicId) === -1) {
+          ancestorTopicIds.push(lastTopicId);
+        }
+
+        unprocessedAncestorTopicIds = unprocessedAncestorTopicIds.concat(
+          this._topicIdToPrerequisiteTopicIds[lastTopicId].filter(
+            (topic) => visitedTopicIdsForCurrentTopic.indexOf(topic) === -1
+          ));
+        visitedTopicIdsForCurrentTopic.push(lastTopicId);
+      }
+      this._topicIdToAncestorTopicIds[topicId] = ancestorTopicIds.sort();
+    }
+  }
+
+  generateTopicIdToSuccessorTopicIds(): void {
+    // The method generates a dict with topic ID as the key and all of its
+    // successor topic IDs as value. The successor's list is generated using
+    // the prerequisite dependency graph.
+    // Example: A -> B -> C, here the topic ID to successor topic IDs dict will
+    // look like: {A: [B, C], B: [C], C: []}.
+    let topicIdToChildTopicId: TopicIdToRelatedTopicIds = {};
```

Comment:
Generating this dict is its own logical bit of functionality, so I'd suggest splitting it out into its own private method.

Reply:
Refactored